File: tokenmaxxing-kde/src/providers/opencode.rs
use crate::creds;
use crate::model::{
    Authority, DayPoint, Gauge, Heatmap, Segment, Snapshot, TokenBreakdown, Totals, Unit, Usage,
    WinStat, Windows,
};
use chrono::{Datelike, Local, NaiveDate};
use rusqlite::{Connection, OpenFlags};
use std::time::Duration;

const CAP_5H: f64 = 12.0;
const CAP_7D: f64 = 30.0;
const CAP_30D: f64 = 60.0;

const NOTE: &str = "No usage API exists. Estimated from this machine's opencode.db against Go's rolling dollar caps — may miss usage on other machines and server-side accounting.";
const PROVIDER: &str = "opencode-go";

pub fn fetch() -> Snapshot {
    if !creds::opencode_go_configured() {
        return degraded("opencode-go not configured — run `opencode auth login`");
    }
    if !creds::opencode_db_path().exists() {
        return degraded("no opencode.db on this machine yet");
    }
    match collect() {
        Ok((gauges, details)) => Snapshot {
            provider_id: PROVIDER.into(),
            provider_name: "opencode go".into(),
            subtitle: "$10/mo · estimated locally".into(),
            authority: Authority::Estimated,
            source: "local opencode.db · estimate".into(),
            gauges,
            details,
            note: Some(NOTE.into()),
            error: None,
            spend: None,
        },
        Err(error) => degraded(&error),
    }
}

/// Full local-history analytics across every provider opencode has run — the
/// paid Go gateway plus any free/local models — not just the capped Go spend.
pub fn usage() -> Usage {
    if !creds::opencode_db_path().exists() {
        return usage_unavailable("no opencode.db on this machine yet");
    }
    match collect_usage() {
        Ok(usage) => usage,
        Err(error) => usage_unavailable(&error),
    }
}

fn collect() -> Result<(Vec<Gauge>, Vec<(String, String)>), String> {
    let connection = open_read_only()?;
    let now_ms = Local::now().timestamp_millis();

    let windows = [
        ("5h", "5-hour rolling", 5 * 3_600_000_i64, CAP_5H),
        ("7d", "Weekly rolling", 7 * 24 * 3_600_000, CAP_7D),
        ("30d", "Monthly rolling", 30 * 24 * 3_600_000, CAP_30D),
    ];
    let gauges = windows
        .into_iter()
        .map(|(key, label, span_ms, cap)| {
            let (spend, requests) = window_stats(&connection, now_ms - span_ms)?;
            Ok(Gauge {
                key: key.into(),
                label: label.into(),
                fraction: (spend / cap).clamp(0.0, 1.0),
                used: Some(spend),
                limit: Some(cap),
                unit: Unit::Usd,
                detail: Some(format!("{requests} req")),
                ..Default::default()
            })
        })
        .collect::<Result<Vec<_>, String>>()?;

    Ok((gauges, all_time_details(&connection)?))
}

fn collect_usage() -> Result<Usage, String> {
    let db = open_read_only()?;

    let daily = daily_series(&db)?;
    let today = Local::now().date_naive().num_days_from_ce();
    let mut windows = Windows::default();
    let mut cost_usd = 0.0;
    for d in &daily {
        cost_usd += d.cost;
        let win = |w: &mut WinStat| {
            w.cost += d.cost;
            w.tokens += d.tokens;
            w.messages += d.messages;
        };
        if d.date_ord == today {
            win(&mut windows.today);
        }
        if d.date_ord > today - 7 {
            win(&mut windows.seven);
        }
        if d.date_ord > today - 30 {
            win(&mut windows.thirty);
        }
    }

    let tokens = token_breakdown(&db)?;
    let (messages, sessions, projects) = counts(&db)?;
    let _ = projects;
    let first = daily.first().map(|d| d.date_ord).and_then(NaiveDate::from_num_days_from_ce_opt);
    let last = daily.last().map(|d| d.date_ord).and_then(NaiveDate::from_num_days_from_ce_opt);

    let totals = Totals {
        cost_usd,
        input: tokens.input,
        output: tokens.output,
        cache_write: tokens.cache_write,
        cache_read: tokens.cache_read,
        messages,
        sessions,
        active_days: daily.len() as u64,
        web_search: 0,
        web_fetch: 0,
        first_day: first,
        last_day: last,
    };

    Ok(Usage {
        scope: "opencode".into(),
        authority: Authority::Estimated,
        source: "local opencode.db · all providers".into(),
        totals,
        windows,
        daily,
        by_model: segments(&db, "json_extract(data,'$.modelID')")?,
        by_project: Vec::new(),
        by_provider: segments(&db, "json_extract(data,'$.providerID')")?,
        tokens,
        heatmap: heatmap(&db)?,
        error: None,
    })
}

/// Open the live opencode SQLite database for reading only. The connection is
/// opened read/write-capable so it can participate in WAL, but `query_only`
/// forbids any mutation of the user's data.
fn open_read_only() -> Result<Connection, String> {
    let path = creds::opencode_db_path();
    let connection = Connection::open_with_flags(&path, OpenFlags::SQLITE_OPEN_READ_WRITE)
        .map_err(|e| format!("open opencode.db: {e}"))?;
    let _ = connection.busy_timeout(Duration::from_millis(2000));
    connection
        .pragma_update(None, "query_only", true)
        .map_err(|e| format!("query_only: {e}"))?;
    Ok(connection)
}

fn window_stats(connection: &Connection, cutoff_ms: i64) -> Result<(f64, i64), String> {
    connection
        .query_row(
            "SELECT COALESCE(SUM(json_extract(data,'$.cost')),0.0), COUNT(*) \
             FROM message \
             WHERE json_extract(data,'$.providerID')=?1 \
               AND json_extract(data,'$.cost') IS NOT NULL \
               AND time_created >= ?2",
            rusqlite::params![PROVIDER, cutoff_ms],
            |row| Ok((row.get::<_, f64>(0)?, row.get::<_, i64>(1)?)),
        )
        .map_err(|e| format!("usage query: {e}"))
}

fn all_time_details(connection: &Connection) -> Result<Vec<(String, String)>, String> {
    let row = connection
        .query_row(
            "SELECT COALESCE(SUM(json_extract(data,'$.cost')),0.0), \
                    COUNT(DISTINCT session_id), \
                    COALESCE(SUM(json_extract(data,'$.tokens.input')),0), \
                    COALESCE(SUM(json_extract(data,'$.tokens.output')),0), \
                    COALESCE(SUM(json_extract(data,'$.tokens.cache.read')),0) \
             FROM message WHERE json_extract(data,'$.providerID')=?1",
            [PROVIDER],
            |row| {
                Ok((
                    row.get::<_, f64>(0)?,
                    row.get::<_, i64>(1)?,
                    row.get::<_, i64>(2)?,
                    row.get::<_, i64>(3)?,
                    row.get::<_, i64>(4)?,
                ))
            },
        )
        .map_err(|e| format!("stats query: {e}"))?;
    let (spend, sessions, input, output, cache_read) = row;
    Ok(vec![
        ("Go spend (all-time)".into(), format!("${spend:.2}")),
        ("Go sessions".into(), sessions.to_string()),
        ("Go tokens in".into(), human_count(input)),
        ("Go tokens out".into(), human_count(output)),
        ("Go cache read".into(), human_count(cache_read)),
    ])
}

/// Cost, tokens and message count per local calendar day, ascending.
fn daily_series(db: &Connection) -> Result<Vec<DayPoint>, String> {
    let mut stmt = db
        .prepare(
            "SELECT date(time_created/1000,'unixepoch','localtime') d, \
                    COALESCE(SUM(json_extract(data,'$.cost')),0.0), \
                    COALESCE(SUM(json_extract(data,'$.tokens.total')),0), \
                    COUNT(*) \
             FROM message WHERE json_extract(data,'$.role')='assistant' \
             GROUP BY d ORDER BY d",
        )
        .map_err(|e| format!("daily prepare: {e}"))?;
    let rows = stmt
        .query_map([], |row| {
            Ok((
                row.get::<_, String>(0)?,
                row.get::<_, f64>(1)?,
                row.get::<_, i64>(2)?,
                row.get::<_, i64>(3)?,
            ))
        })
        .map_err(|e| format!("daily query: {e}"))?;
    let mut out = Vec::new();
    for row in rows {
        let (date, cost, tokens, msgs) = row.map_err(|e| e.to_string())?;
        let Ok(nd) = NaiveDate::parse_from_str(&date, "%Y-%m-%d") else { continue };
        out.push(DayPoint {
            date_ord: nd.num_days_from_ce(),
            cost,
            tokens: tokens.max(0) as u64,
            messages: msgs.max(0) as u64,
        });
    }
    Ok(out)
}

fn token_breakdown(db: &Connection) -> Result<TokenBreakdown, String> {
    db.query_row(
        "SELECT COALESCE(SUM(json_extract(data,'$.tokens.input')),0), \
                COALESCE(SUM(json_extract(data,'$.tokens.output')),0), \
                COALESCE(SUM(json_extract(data,'$.tokens.cache.write')),0), \
                COALESCE(SUM(json_extract(data,'$.tokens.cache.read')),0), \
                COALESCE(SUM(json_extract(data,'$.tokens.reasoning')),0) \
         FROM message WHERE json_extract(data,'$.role')='assistant'",
        [],
        |row| {
            Ok(TokenBreakdown {
                input: row.get::<_, i64>(0)?.max(0) as u64,
                output: row.get::<_, i64>(1)?.max(0) as u64,
                cache_write: row.get::<_, i64>(2)?.max(0) as u64,
                cache_read: row.get::<_, i64>(3)?.max(0) as u64,
                reasoning: row.get::<_, i64>(4)?.max(0) as u64,
            })
        },
    )
    .map_err(|e| format!("tokens query: {e}"))
}

fn counts(db: &Connection) -> Result<(u64, u64, u64), String> {
    let messages = db
        .query_row(
            "SELECT COUNT(*) FROM message WHERE json_extract(data,'$.role')='assistant'",
            [],
            |r| r.get::<_, i64>(0),
        )
        .map_err(|e| format!("msg count: {e}"))?;
    let sessions = db
        .query_row("SELECT COUNT(*) FROM session", [], |r| r.get::<_, i64>(0))
        .unwrap_or(0);
    let projects = db
        .query_row("SELECT COUNT(*) FROM project", [], |r| r.get::<_, i64>(0))
        .unwrap_or(0);
    Ok((messages.max(0) as u64, sessions.max(0) as u64, projects.max(0) as u64))
}

/// Cost / token / message totals grouped by an arbitrary JSON expression
/// (provider id or model id), sorted by tokens desc.
fn segments(db: &Connection, group_expr: &str) -> Result<Vec<Segment>, String> {
    let sql = format!(
        "SELECT COALESCE({group_expr},'?') g, \
                COALESCE(SUM(json_extract(data,'$.cost')),0.0), \
                COALESCE(SUM(json_extract(data,'$.tokens.total')),0), \
                COUNT(*) \
         FROM message WHERE json_extract(data,'$.role')='assistant' \
         GROUP BY g ORDER BY 3 DESC"
    );
    let mut stmt = db.prepare(&sql).map_err(|e| format!("segment prepare: {e}"))?;
    let rows = stmt
        .query_map([], |row| {
            Ok(Segment {
                label: row.get::<_, String>(0)?,
                cost: row.get::<_, f64>(1)?,
                tokens: row.get::<_, i64>(2)?.max(0) as u64,
                messages: row.get::<_, i64>(3)?.max(0) as u64,
            })
        })
        .map_err(|e| format!("segment query: {e}"))?;
    rows.collect::<Result<Vec<_>, _>>().map_err(|e| e.to_string())
}

fn heatmap(db: &Connection) -> Result<Heatmap, String> {
    let mut stmt = db
        .prepare(
            "SELECT CAST(strftime('%w', time_created/1000,'unixepoch','localtime') AS INTEGER), \
                    CAST(strftime('%H', time_created/1000,'unixepoch','localtime') AS INTEGER), \
                    COUNT(*) \
             FROM message WHERE json_extract(data,'$.role')='assistant' \
             GROUP BY 1, 2",
        )
        .map_err(|e| format!("heatmap prepare: {e}"))?;
    let rows = stmt
        .query_map([], |row| {
            Ok((row.get::<_, i64>(0)?, row.get::<_, i64>(1)?, row.get::<_, i64>(2)?))
        })
        .map_err(|e| format!("heatmap query: {e}"))?;
    let mut heat = Heatmap::default();
    for row in rows {
        let (sun_dow, hour, count) = row.map_err(|e| e.to_string())?;
        let weekday = ((sun_dow + 6) % 7) as usize; // strftime 0=Sun → Monday-based
        let hour = hour.clamp(0, 23) as usize;
        let count = count.max(0) as u64;
        heat.counts[weekday][hour] += count;
        heat.max = heat.max.max(heat.counts[weekday][hour]);
    }
    Ok(heat)
}

fn human_count(n: i64) -> String {
    let value = n as f64;
    if value >= 1e9 {
        format!("{:.1}B", value / 1e9)
    } else if value >= 1e6 {
        format!("{:.1}M", value / 1e6)
    } else if value >= 1e3 {
        format!("{:.1}K", value / 1e3)
    } else {
        n.to_string()
    }
}

fn degraded(message: &str) -> Snapshot {
    Snapshot {
        provider_id: PROVIDER.into(),
        provider_name: "opencode go".into(),
        subtitle: "$10/mo subscription".into(),
        authority: Authority::Unavailable,
        source: "local opencode.db · unavailable".into(),
        gauges: Vec::new(),
        details: Vec::new(),
        note: None,
        error: Some(message.to_string()),
        spend: None,
    }
}

fn usage_unavailable(message: &str) -> Usage {
    Usage {
        scope: "opencode".into(),
        authority: Authority::Unavailable,
        source: "local opencode.db · unavailable".into(),
        error: Some(message.to_string()),
        ..Default::default()
    }
}

File: tokenmaxxing-macos/Sources/Tokenmaxxing/Providers/OpenCode.swift
import Foundation
import SQLite3

private let SQLITE_TRANSIENT = unsafeBitCast(-1, to: sqlite3_destructor_type.self)

struct OpenCodeAllTime {
    var spend = 0.0
    var sessions: Int64 = 0
    var tokensIn: Int64 = 0
    var tokensOut: Int64 = 0
    var cacheRead: Int64 = 0
}

/// opencode-go has no usage API, so this estimates rolling-window spend from the
/// local opencode.db against Go's dollar caps, and separately aggregates full
/// usage history across every provider opencode has run.
enum OpenCodeProvider {
    static let cap5h = 12.0
    static let cap7d = 30.0
    static let cap30d = 60.0
    static let provider = "opencode-go"

    static func fetch() -> Snapshot {
        guard Creds.opencodeGoConfigured() else {
            return degraded("opencode-go not configured — run `opencode auth login`")
        }
        guard FileManager.default.fileExists(atPath: Creds.opencodeDbPath.path) else {
            return degraded("no opencode.db on this machine yet")
        }
        do {
            let (gauges, details, remote) = try collect()
            return Snapshot(
                providerId: provider,
                providerName: "opencode go",
                subtitle: subtitle(for: remote),
                authority: .estimated,
                source: source(for: remote),
                gauges: gauges,
                details: details,
                note: note(for: remote),
                error: nil
            )
        } catch {
            return degraded("\(error)")
        }
    }

    /// Full local-history analytics across every provider opencode has run — the
    /// paid Go gateway plus any free/local models — not just the capped Go spend.
    static func usage() -> Usage {
        guard FileManager.default.fileExists(atPath: Creds.opencodeDbPath.path) else {
            return usageUnavailable("no opencode.db on this machine yet")
        }
        do {
            let db = try Database(path: Creds.opencodeDbPath.path)
            defer { db.close() }
            let daily = try db.dailySeries()

            let calendar = Calendar.current
            let today = calendar.startOfDay(for: Date())
            let cutoff7 = calendar.date(byAdding: .day, value: -6, to: today)!
            let cutoff30 = calendar.date(byAdding: .day, value: -29, to: today)!
            var windows = Windows()
            var cost = 0.0
            for d in daily {
                cost += d.cost
                if d.date == today { fold(&windows.today, d) }
                if d.date >= cutoff7 { fold(&windows.seven, d) }
                if d.date >= cutoff30 { fold(&windows.thirty, d) }
            }

            let tokens = try db.tokenBreakdown()
            let (messages, sessions) = try db.counts()
            var totals = Totals()
            totals.costUSD = cost
            totals.input = tokens.input
            totals.output = tokens.output
            totals.cacheWrite = tokens.cacheWrite
            totals.cacheRead = tokens.cacheRead
            totals.messages = messages
            totals.sessions = sessions
            totals.activeDays = Int64(daily.count)
            totals.firstDay = daily.first?.date
            totals.lastDay = daily.last?.date

            return Usage(
                scope: "opencode",
                authority: .estimated,
                source: "local opencode.db · all providers",
                totals: totals,
                windows: windows,
                daily: daily,
                byModel: try db.segments(groupExpr: "json_extract(data,'$.modelID')"),
                byProject: [],
                byProvider: try db.segments(groupExpr: "json_extract(data,'$.providerID')"),
                tokens: tokens,
                heatmap: try db.heatmap(),
                error: nil
            )
        } catch {
            return usageUnavailable("\(error)")
        }
    }

    private static func fold(_ w: inout WinStat, _ d: DayPoint) {
        w.cost += d.cost
        w.tokens += d.tokens
        w.messages += d.messages
    }

    private static func collect() throws -> ([Gauge], [Detail], OpenCodeRemoteReport) {
        let db = try Database(path: Creds.opencodeDbPath.path)
        defer { db.close() }
        let remote = OpenCodeRemote.report()
        let nowMs = Int64(Date().timeIntervalSince1970 * 1000)
        let windows: [(String, String, Int64, Double)] = [
            ("5h", "5-hour rolling", 5 * 3_600_000, cap5h),
            ("7d", "Weekly rolling", 7 * 24 * 3_600_000, cap7d),
            ("30d", "Monthly rolling", 30 * 24 * 3_600_000, cap30d),
        ]
        var gauges: [Gauge] = []
        for (key, label, span, cap) in windows {
            let (localSpend, localRequests) = try db.windowStats(cutoffMs: nowMs - span, provider: provider)
            let extra = remote.stats.windows[key] ?? OpenCodeRemoteWindow()
            let spend = localSpend + extra.spend
            let requests = localRequests + extra.requests
            gauges.append(Gauge(
                key: key, label: label,
                fraction: min(1, max(0, spend / cap)),
                used: spend, limit: cap, unit: .usd,
                detail: "\(requests) req", resetsAt: nil, trustedReset: false
            ))
        }
        let details = detailRows(local: try db.allTimeStats(provider: provider), remote: remote)
        return (gauges, details, remote)
    }

    private static func detailRows(local: OpenCodeAllTime, remote: OpenCodeRemoteReport) -> [Detail] {
        var rows: [Detail] = []
        if !remote.configured.isEmpty {
            rows.append(Detail(key: "Machines", value: coverage(remote)))
        }
        rows += [
            Detail(key: "Go spend (all-time)", value: String(format: "$%.2f", local.spend + remote.stats.allTimeSpend)),
            Detail(key: "Go sessions", value: "\(local.sessions + remote.stats.sessions)"),
            Detail(key: "Go tokens in", value: humanCount(local.tokensIn + remote.stats.tokensIn)),
            Detail(key: "Go tokens out", value: humanCount(local.tokensOut + remote.stats.tokensOut)),
            Detail(key: "Go cache read", value: humanCount(local.cacheRead + remote.stats.cacheRead)),
        ]
        return rows
    }

    private static func coverage(_ remote: OpenCodeRemoteReport) -> String {
        var parts = ["local"]
        parts += remote.reached
        parts += remote.stale.map { "\($0) (cached)" }
        var text = parts.joined(separator: " + ")
        if !remote.unreachable.isEmpty {
            text += " · \(remote.unreachable.joined(separator: ", ")) unreachable"
        }
        return text
    }

    private static func subtitle(for remote: OpenCodeRemoteReport) -> String {
        let machines = 1 + remote.included.count
        return machines > 1 ? "$10/mo · estimated · \(machines) machines" : "$10/mo · estimated locally"
    }

    private static func source(for remote: OpenCodeRemoteReport) -> String {
        guard !remote.included.isEmpty else { return "local opencode.db · estimate" }
        return "opencode.db: \(coverage(remote)) · estimate"
    }

    private static func note(for remote: OpenCodeRemoteReport) -> String {
        guard !remote.configured.isEmpty else {
            return "No usage API exists. Estimated from this machine's opencode.db against Go's rolling dollar caps — may miss usage on other machines and server-side accounting. List other machines in ~/.config/tokenmaxxing/config.json under opencode_remote_hosts to include them over SSH."
        }
        var text = "No usage API exists. Summed from opencode.db on this machine plus \(remote.configured.joined(separator: ", ")) over SSH against Go's rolling dollar caps — may still miss unlisted machines and server-side accounting."
        for host in remote.stale {
            text += " \(host): using figures cached under 15 minutes ago."
        }
        for host in remote.unreachable {
            text += " \(host) unreachable — its spend is not included."
        }
        return text
    }

    private static func degraded(_ message: String) -> Snapshot {
        Snapshot(
            providerId: provider,
            providerName: "opencode go",
            subtitle: "$10/mo subscription",
            authority: .unavailable,
            source: "local opencode.db · unavailable",
            gauges: [],
            details: [],
            note: nil,
            error: message
        )
    }

    private static func usageUnavailable(_ message: String) -> Usage {
        Usage(scope: "opencode", authority: .unavailable, source: "local opencode.db · unavailable", error: message)
    }
}

/// Minimal read-only SQLite wrapper. Opened read/write so it can participate in
/// WAL, but `query_only` forbids any mutation of the user's data.
private final class Database {
    let handle: OpaquePointer

    private let dayFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        f.timeZone = .current
        f.locale = Locale(identifier: "en_US_POSIX")
        return f
    }()

    init(path: String) throws {
        var db: OpaquePointer?
        guard sqlite3_open_v2(path, &db, SQLITE_OPEN_READWRITE, nil) == SQLITE_OK, let db else {
            throw QuotaError.message("open opencode.db failed")
        }
        handle = db
        sqlite3_busy_timeout(db, 2000)
        sqlite3_exec(db, "PRAGMA query_only=ON;", nil, nil, nil)
    }

    func close() {
        sqlite3_close(handle)
    }

    func windowStats(cutoffMs: Int64, provider: String) throws -> (Double, Int64) {
        let sql = """
        SELECT COALESCE(SUM(json_extract(data,'$.cost')),0.0), COUNT(*)
        FROM message
        WHERE json_extract(data,'$.providerID')=?1
          AND json_extract(data,'$.cost') IS NOT NULL
          AND time_created >= ?2
        """
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(handle, sql, -1, &stmt, nil) == SQLITE_OK else {
            throw QuotaError.message("usage query prepare failed")
        }
        defer { sqlite3_finalize(stmt) }
        sqlite3_bind_text(stmt, 1, provider, -1, SQLITE_TRANSIENT)
        sqlite3_bind_int64(stmt, 2, cutoffMs)
        guard sqlite3_step(stmt) == SQLITE_ROW else { return (0, 0) }
        return (sqlite3_column_double(stmt, 0), sqlite3_column_int64(stmt, 1))
    }

    func allTimeStats(provider: String) throws -> OpenCodeAllTime {
        let sql = """
        SELECT COALESCE(SUM(json_extract(data,'$.cost')),0.0),
               COUNT(DISTINCT session_id),
               COALESCE(SUM(json_extract(data,'$.tokens.input')),0),
               COALESCE(SUM(json_extract(data,'$.tokens.output')),0),
               COALESCE(SUM(json_extract(data,'$.tokens.cache.read')),0)
        FROM message WHERE json_extract(data,'$.providerID')=?1
        """
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(handle, sql, -1, &stmt, nil) == SQLITE_OK else {
            throw QuotaError.message("stats query prepare failed")
        }
        defer { sqlite3_finalize(stmt) }
        sqlite3_bind_text(stmt, 1, provider, -1, SQLITE_TRANSIENT)
        guard sqlite3_step(stmt) == SQLITE_ROW else { return OpenCodeAllTime() }
        return OpenCodeAllTime(
            spend: sqlite3_column_double(stmt, 0),
            sessions: sqlite3_column_int64(stmt, 1),
            tokensIn: sqlite3_column_int64(stmt, 2),
            tokensOut: sqlite3_column_int64(stmt, 3),
            cacheRead: sqlite3_column_int64(stmt, 4)
        )
    }

    func dailySeries() throws -> [DayPoint] {
        let sql = """
        SELECT date(time_created/1000,'unixepoch','localtime') d,
               COALESCE(SUM(json_extract(data,'$.cost')),0.0),
               COALESCE(SUM(json_extract(data,'$.tokens.total')),0),
               COUNT(*)
        FROM message WHERE json_extract(data,'$.role')='assistant'
        GROUP BY d ORDER BY d
        """
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(handle, sql, -1, &stmt, nil) == SQLITE_OK else {
            throw QuotaError.message("daily query prepare failed")
        }
        defer { sqlite3_finalize(stmt) }
        var out: [DayPoint] = []
        while sqlite3_step(stmt) == SQLITE_ROW {
            guard let cstr = sqlite3_column_text(stmt, 0) else { continue }
            guard let date = dayFormatter.date(from: String(cString: cstr)) else { continue }
            out.append(DayPoint(
                date: Calendar.current.startOfDay(for: date),
                cost: sqlite3_column_double(stmt, 1),
                tokens: max(0, sqlite3_column_int64(stmt, 2)),
                messages: max(0, sqlite3_column_int64(stmt, 3))
            ))
        }
        return out
    }

    func tokenBreakdown() throws -> TokenBreakdown {
        let sql = """
        SELECT COALESCE(SUM(json_extract(data,'$.tokens.input')),0),
               COALESCE(SUM(json_extract(data,'$.tokens.output')),0),
               COALESCE(SUM(json_extract(data,'$.tokens.cache.write')),0),
               COALESCE(SUM(json_extract(data,'$.tokens.cache.read')),0),
               COALESCE(SUM(json_extract(data,'$.tokens.reasoning')),0)
        FROM message WHERE json_extract(data,'$.role')='assistant'
        """
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(handle, sql, -1, &stmt, nil) == SQLITE_OK else {
            throw QuotaError.message("tokens query prepare failed")
        }
        defer { sqlite3_finalize(stmt) }
        guard sqlite3_step(stmt) == SQLITE_ROW else { return TokenBreakdown() }
        return TokenBreakdown(
            input: max(0, sqlite3_column_int64(stmt, 0)),
            output: max(0, sqlite3_column_int64(stmt, 1)),
            cacheWrite: max(0, sqlite3_column_int64(stmt, 2)),
            cacheRead: max(0, sqlite3_column_int64(stmt, 3)),
            reasoning: max(0, sqlite3_column_int64(stmt, 4))
        )
    }

    func counts() throws -> (Int64, Int64) {
        let messages = scalar("SELECT COUNT(*) FROM message WHERE json_extract(data,'$.role')='assistant'")
        let sessions = scalar("SELECT COUNT(*) FROM session")
        return (messages, sessions)
    }

    private func scalar(_ sql: String) -> Int64 {
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(handle, sql, -1, &stmt, nil) == SQLITE_OK else { return 0 }
        defer { sqlite3_finalize(stmt) }
        guard sqlite3_step(stmt) == SQLITE_ROW else { return 0 }
        return max(0, sqlite3_column_int64(stmt, 0))
    }

    /// Cost / token / message totals grouped by a JSON expression, sorted by tokens desc.
    func segments(groupExpr: String) throws -> [Segment] {
        let sql = """
        SELECT COALESCE(\(groupExpr),'?') g,
               COALESCE(SUM(json_extract(data,'$.cost')),0.0),
               COALESCE(SUM(json_extract(data,'$.tokens.total')),0),
               COUNT(*)
        FROM message WHERE json_extract(data,'$.role')='assistant'
        GROUP BY g ORDER BY 3 DESC
        """
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(handle, sql, -1, &stmt, nil) == SQLITE_OK else {
            throw QuotaError.message("segment query prepare failed")
        }
        defer { sqlite3_finalize(stmt) }
        var out: [Segment] = []
        while sqlite3_step(stmt) == SQLITE_ROW {
            let label = sqlite3_column_text(stmt, 0).map { String(cString: $0) } ?? "?"
            out.append(Segment(
                label: label,
                cost: sqlite3_column_double(stmt, 1),
                tokens: max(0, sqlite3_column_int64(stmt, 2)),
                messages: max(0, sqlite3_column_int64(stmt, 3))
            ))
        }
        return out
    }

    func heatmap() throws -> Heatmap {
        let sql = """
        SELECT CAST(strftime('%w', time_created/1000,'unixepoch','localtime') AS INTEGER),
               CAST(strftime('%H', time_created/1000,'unixepoch','localtime') AS INTEGER),
               COUNT(*)
        FROM message WHERE json_extract(data,'$.role')='assistant'
        GROUP BY 1, 2
        """
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(handle, sql, -1, &stmt, nil) == SQLITE_OK else {
            throw QuotaError.message("heatmap query prepare failed")
        }
        defer { sqlite3_finalize(stmt) }
        var heat = Heatmap()
        while sqlite3_step(stmt) == SQLITE_ROW {
            let sunDow = Int(sqlite3_column_int64(stmt, 0))
            let hour = Int(sqlite3_column_int64(stmt, 1))
            let count = Int(max(0, sqlite3_column_int64(stmt, 2)))
            let weekday = (sunDow + 6) % 7 // strftime 0=Sun → Monday-based
            guard (0..<7).contains(weekday), (0..<24).contains(hour) else { continue }
            heat.counts[weekday][hour] += count
            heat.max = max(heat.max, heat.counts[weekday][hour])
        }
        return heat
    }
}

private func humanCount(_ n: Int64) -> String {
    let value = Double(n)
    if value >= 1e9 { return String(format: "%.1fB", value / 1e9) }
    if value >= 1e6 { return String(format: "%.1fM", value / 1e6) }
    if value >= 1e3 { return String(format: "%.1fK", value / 1e3) }
    return "\(n)"
}
